### PR TITLE
Manage wasm invalid grpc code

### PR DESCRIPTION
**Pull request by juanmolle**
Commit Message: wasm: Manage wasm invalid grpc code

Additional Description: In the scenario when the wasm generate a local reply with unset (invalid) grpc code. instead of propagate the -1, now the context will check and propagate the nullopt to let envoy populate it correctly.

Risk Level: Low
Testing: yes
Docs Changes: no
Release Notes: no
Platform Specific Features: n/a

### CODE REVIEW

**Comment by juanmolle** [comment]
Do I need to add this into the release not?

**Comment by juanmolle** [comment]
cc @PiotrSikora

File: source/extensions/common/wasm/context.cc [hunk 1]
```diff
@@ -1691,8 +1691,14 @@ WasmResult Context::sendLocalResponse(uint32_t response_code, std::string_view b
       if (local_reply_sent_) {
         return;
       }
+      // In case -1 is send from sdk, as an invalid value, nullopt is sent to allows envoy to
+      // set the grpc status code
```
Comment (PiotrSikora) [inline source/extensions/common/wasm/context.cc:1695]: I'd rephrase it slightly to mention that C++ and Rust SDKs (and maybe others?) use `-1` (`InvalidCode`) as the default value in case gRPC status is not set, which should be mapped to `nullopt` in Envoy to prevent it from sending `grpc-status` trailer at all.

**Review by PiotrSikora — COMMENTED** [review]
Thanks! This is a nice targeted fix, but this issue isn't unique to Wasm, and I think that this should be fixed in `sendLocalReply` to handle `-1`, since it's unlikely that anybody wants to send it as `grpc-status` value.

**Comment by juanmolle** [comment]
@PiotrSikora  do you mean to just add a check here https://github.com/envoyproxy/envoy/blob/main/source/common/http/utility.cc#L697

```
    if (response_headers->getGrpcStatusValue().empty()) {
      response_headers->setGrpcStatus(std::to_string(
          enumToInt(local_reply_data.grpc_status_ 
                        ? local_reply_data.grpc_status_.value()
                        : Grpc::Utility::httpToGrpcStatus(enumToInt(response_code)))));
    }
```
To this?

```
    if (response_headers->getGrpcStatusValue().empty()) {
      response_headers->setGrpcStatus(std::to_string(
          enumToInt(local_reply_data.grpc_status_ &&
                    local_reply_data.grpc_status_.value() != Grpc::Status::WellKnownGrpcStatus::InvalidCode
                        ? local_reply_data.grpc_status_.value()
                        : Grpc::Utility::httpToGrpcStatus(enumToInt(response_code)))));
    }
```

Should I keep the check in the wasm context? I think sanitizing it as soon as possible is ok. But I could let it go if you prefer.

**Comment by juanmolle** [comment]
@https://github.com/envoyproxy/envoy/pull/33867

File: source/extensions/common/wasm/context.cc [hunk 2]
```diff
@@ -1691,8 +1691,17 @@ WasmResult Context::sendLocalResponse(uint32_t response_code, std::string_view b
       if (local_reply_sent_) {
         return;
       }
+      // C++, Rust and other SDKs use -1 (InvalidCode) as the default value if gRPC code is not set,
+      // which should be mapped to nullopt in Envoy to prevent it from sending a grpc-status trailer
+      // at all.
+      absl::optional<Grpc::Status::GrpcStatus> grpc_status_code =
+          Grpc::Status::WellKnownGrpcStatus(grpc_status) ==
```
Comment (kyessenov) [inline source/extensions/common/wasm/context.cc:1698]: This is undefined behavior I think for invalid values, e.g. 20? Can we change this to:
```c++
abs::optional<...> grpc_status_code = absl::nullopt;
if (grpc_status >= Ok && grpc_status <= MaximumKnown) {
  grpc_status_code = Grpc::Status::WellKnownGrpcStatus(grpc_status);
}
```

**Review by kyessenov — CHANGES_REQUESTED** [review]

File: source/extensions/common/wasm/context.cc (hunk 2 above)
Comment (juanmolle) [reply source/extensions/common/wasm/context.cc:1698]: Make sense, thanks

**Review by kyessenov — APPROVED** [review]

**Comment by kyessenov** [comment]
@PiotrSikora - are you good with this? It would be good to harden the internals of Envoy as well but I think the trust boundary is different for Wasm.

**Comment by PiotrSikora** [comment]
> @PiotrSikora - are you good with this? It would be good to harden the internals of Envoy as well but I think the trust boundary is different for Wasm.

As I've mentioned before, I think that this should be fixed deeper at the Envoy's API level, otherwise other consumers (presumably, including Lua and ext_proc) can send garbage in `grpc-status` header, and the behavior won't be consistent across various extension points, but that's up to the maintainers to decide...

From the Proxy-Wasm point of view, this is fine, since it fixes the issue at hand, so LGTM.

**Comment by kyessenov** [comment]
@PiotrSikora understood, but native extensions are not supposed to send garbage in the call to `sendLocalReply`. Wasm plugins can, since SDKs are optional, so we need extra protection in any case.

File: source/extensions/common/wasm/context.cc (hunk 2 above)
Comment (juanmolle) [reply source/extensions/common/wasm/context.cc]: This is interesting and it was interesting due to this enum is unscoped enum and the underlying type must contains it values, it must be at lease signed char

 > std 7.2.7
 For an enumeration whose underlying type is not fixed, the underlying type is an integral type that can
represent all the enumerator values defined in the enumeration.

then 20 can be assigned to the enumeration. 

> std 7.2.8 
It is possible to define an enumeration that has values not defined by any of its
enumerators. If the enumerator-list is empty, the values of the enumeration are as if the enumeration had a
single enumerator with value 0.

but only if it fits in the underlying type.

anyway it was my bad to cast an uint32 into an enum without fixed underlying type, specially one that is unsigned into an enum that should be signed to hold the -1 value. 

I should have done what you are proposing in the first place, Thanks again, and it is always good to review the standard.
